source/common/stats/thread_local_store.cc

```diff
@@ -290,19 +301,24 @@ Histogram& ThreadLocalStoreImpl::ScopeImpl::histogram(const std::string& name) {
   // share this code so I'm leaving it largely duplicated for now.
   std::string final_name = prefix_ + name;
   ParentHistogramSharedPtr* tls_ref = nullptr;
-
+  Thread::LockGuard lock(parent_.lock_);
+  StatNamePtr stat_name_ptr = parent_.heap_allocator_.encode(final_name);
   if (!parent_.shutting_down_ && parent_.tls_) {
     tls_ref = &parent_.tls_->getTyped<TlsCache>()
                    .scope_cache_[this->scope_id_]
-                   .parent_histograms_[final_name];
+                   .parent_histograms_[std::move(stat_name_ptr)];
   }
 
   if (tls_ref && *tls_ref) {
     return **tls_ref;
   }
 
-  Thread::LockGuard lock(parent_.lock_);
-  ParentHistogramImplSharedPtr& central_ref = central_cache_.histograms_[final_name];
+  // TODO(ambuc): This is obviously a duplicate of the work done just a few lines ago. If these were
```

Comment:
I think shared_ptrs would be a good option here.  A copy factory could also work, but it will have more computational overhead while saving very little memory on average IIUC.  I think a benchmark would be worthwhile to verify that intuition.

Reply:
Yeah I think I like shared pointers here as well...seems like the same name may be referenced more than once in a few places, and it would be nice not to pay the encode/decode each time. IMO this could be a follow-up though.

Reply:
In my experiments, using a shared ptr for StatNamePtr ate into the heap benefits due to the cost of reference counting each stat name (essentially) twice. I'm going to try to write a copy constructor which eases the cost of reencoding instead.

Reply:
@jmarantz @mrice32 I'll leave the shared_ptr aspect to another PR.

Reply:
Are you sure it's better not to ref-count? It looks like what you have will, most of the time, create two distinct equivalent vector<symbol> for each stat. That surely is more expensive than a ref-count.